docs/design/storage_abstraction.md

```diff
@@ -0,0 +1,275 @@
+    public fun borrow_mut<V: key>(table: &mut TypeTable): &mut V;
+
+    #[private_generics(T)]
+    public fun remove<V: key>(table: &mut TypeTable): V;
+
+    #[private_generics(T)]
+    public fun contains<V: key>(table: &TypeTable): bool;
+}
+```
+
+The methods in TypeTable rely on the type safety guarantee of `private_generics`, with constraints similar to Move's global storage instructions.
+
+### AccountStorage
+
+AccountStorage is an abstraction of a user's storage space in Move, which contains two Tables, the Resource Table and Module Table, allowing Resources and Modules to be directly manipulated in Move without the need for global storage instructions.
+
+```move
+struct AccountStorage has key {
+    resources: TypeTable,
+    modules: Table<String, vector<u8>>,
+}
+```
+
+It mainly provides the following API:
+
+```
+module moveos_std::account_storage{
+
+    #[private_generics(T)]
+    /// Move a resource to the account's storage
+    /// This function equates to `move_to<T>(&signer, resource)` instruction in Move
+    public fun global_move_to<T: key>(ctx: &mut StorageContext, account: &signer, resource: T);
+
+    #[private_generics(T)]
+    /// Move a resource from the account's storage
+    /// This function equates to `move_from<T>(address)` instruction in Move
+    public fun global_move_from<T: key>(ctx: &mut StorageContext, account: address): T;
+
+    #[private_generics(T)]
+    /// Check if the account has a resource of the given type
+    /// This function equates to `exists<T>(address)` instruction in Move
+    public fun global_exists<T: key>(ctx: &StorageContext, account: address) : bool;
+
+    /// Check if the account has a module with the given name
+    public fun exists_module(ctx: &StorageContext, account: address, name: String): bool;
+
+    /// Publish modules to the account's storage
+    public fun publish_modules(ctx: &mut StorageContext, account: &signer, modules: vector<vector<u8>>);
+}
+```
+
+The above methods can replace the global storage instructions in Move, so that all storage-related operations in Move are ultimately unified to Operations on Tables. 
+
+At the same time,this approach also solves a reference problem in Move whereby developers cannot return a reference borrowed through the `borrow_global<T>` instruction in their methods, since this reference is created out of thin air and cannot be returned as a method return value. However, if the reference is borrowed through `StorageContext`, then that goal can be achieved, and the lifetime of the reference is bound to the lifetime of `StorageContext`. This eliminates the need for developers to manually track reference lifetimes and ensures the safety of their usage in the Move program.
+
+In addition, AccountStorage also provides module-related methods, making it easy to deploy Move contracts in Move and allowing developers to use contracts to define contract deployment rules, such as upgrading the contract through DAO governance.
```

Comment:
what's the method for updating a module? (i.e. what happens if I publish a module with an existed name) and how exactly do I define the deployment rules?
Maybe adding an example here?

Reply:
I provide a `publish_modules` function, but it is incomplete. I write two issues to track this #83 #84